### ISSUE TITLE
Valgrind error: `invalid write of size 8` in backfill `chunk_callback_t::coro_pool_callback`

### ISSUE THREAD

**Issue by Tryneus** [rethinkdb/rethinkdb#3591]
This showed up when running the shard fuzzer with valgrind on `next` overnight:

```
==23436== Invalid write of size 8
==23436==    at 0x1D15206: chunk_callback_t::coro_pool_callback(backfill_queue_entry_t, signal_t*) (backfillee.cc:93)
==23436==    by 0x1D19EE9: coro_pool_t<backfill_queue_entry_t>::worker_run(backfill_queue_entry_t, auto_drainer_t::lock_t) (coro_pool.hpp:68)
==23436==    by 0x1D1D3AD: std::_Mem_fn<void (coro_pool_t<backfill_queue_entry_t>::*)(backfill_queue_entry_t, auto_drainer_t::lock_t)>::operator()(coro_pool_t<backfill_queue_entry_t>*, backfill_queue_entry_t, auto_drainer_t::lock_t) const (in /home/ssd2/grey/rethinkdb/build/debug_valgrind_system/rethinkdb)
==23436==    by 0x1D1D033: _ZNSt5_BindIFSt7_Mem_fnIM11coro_pool_tI22backfill_queue_entry_tEFvS2_N14auto_drainer_t6lock_tEEEPS3_S2_S5_EE6__callIvIEILi0ELi1ELi2EEEET_OSt5tupleIIDpT0_EESt12_Index_tupleIIXspT1_EEE (functional:1146)
==23436==    by 0x1D1CA21: void std::_Bind<std::_Mem_fn<void (coro_pool_t<backfill_queue_entry_t>::*)(backfill_queue_entry_t, auto_drainer_t::lock_t)> ()(coro_pool_t<backfill_queue_entry_t>*, backfill_queue_entry_t, auto_drainer_t::lock_t)>::operator()<, void>() (functional:1205)
==23436==    by 0x1D1C0DB: callable_action_instance_t<std::_Bind<std::_Mem_fn<void (coro_pool_t<backfill_queue_entry_t>::*)(backfill_queue_entry_t, auto_drainer_t::lock_t)> ()(coro_pool_t<backfill_queue_entry_t>*, backfill_queue_entry_t, auto_drainer_t::lock_t)> >::run_action() (callable_action.hpp:31)
==23436==    by 0x1590677: callable_action_wrapper_t::run() (runtime_utils.cc:42)
==23436==    by 0x1580392: coro_t::run() (coroutines.cc:196)
==23436==  Address 0x1b73ff00 is 16 bytes inside a block of size 32 free'd
==23436==    at 0x4C2A4BC: operator delete(void*) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==23436==    by 0x1605D64: void clear_and_delete<buf_lock_t>(std::vector<buf_lock_t*, std::allocator<buf_lock_t*> >*) (blob.cc:25)
==23436==    by 0x1602FBA: blob_acq_t::~blob_acq_t() (blob.cc:32)
==23436==    by 0x16046DA: blob_t::write_from_string(std::string const&, buf_parent_t, long) (blob.cc:560)
==23436==    by 0x1BB8927: void metadata_persistence::write_blob<(cluster_version_t)4, branch_history_t>(buf_parent_t, char*, int, branch_history_t const&) (persist.cc:129)
==23436==    by 0x1BBAECC: void metadata_persistence::write_branch_history_blob<(cluster_version_t)4>(buf_parent_t, metadata_persistence::cluster_metadata_superblock_t*, branch_history_t const&) (persist.cc:244)
==23436==    by 0x1BBA4BD: metadata_persistence::cluster_persistent_file_t::persistent_branch_history_manager_t::flush() (persist.cc:630)
==23436==    by 0x1BC1EAF: std::_Mem_fn<void (metadata_persistence::cluster_persistent_file_t::persistent_branch_history_manager_t::*)()>::operator()(metadata_persistence::cluster_persistent_file_t::persistent_branch_history_manager_t*) const (in /home/ssd2/grey/rethinkdb/build/debug_valgrind_system/rethinkdb)
==23436==    by 0x1BC12FD: _ZNSt5_BindIFSt7_Mem_fnIMN20metadata_persistence25cluster_persistent_file_t35persistent_branch_history_manager_tEFvvEEPS3_EE6__callIvIEILi0EEEET_OSt5tupleIIDpT0_EESt12_Index_tupleIIXspT1_EEE (functional:1146)
==23436==    by 0x1BBF933: void std::_Bind<std::_Mem_fn<void (metadata_persistence::cluster_persistent_file_t::persistent_branch_history_manager_t::*)()> ()(metadata_persistence::cluster_persistent_file_t::persistent_branch_history_manager_t*)>::operator()<, void>() (functional:1205)
==23436==    by 0x1BBDF17: std::_Function_handler<void ()(), std::_Bind<std::_Mem_fn<void (metadata_persistence::cluster_persistent_file_t::persistent_branch_history_manager_t::*)()> ()(metadata_persistence::cluster_persistent_file_t::persistent_branch_history_manager_t*)> >::_M_invoke(std::_Any_data const&) (functional:1778)
==23436==    by 0x1576A8F: std::function<void ()()>::operator()() const (functional:2161)
```

**Comment by Tryneus** [comment]
This may possibly be the cause of #3576 and #3577.  Going to fix this first and see if the other problems persist.

**Comment by Tryneus** [comment]
A similar error from a different server, with a different deallocation trace:

```
==23122== Invalid write of size 8
==23122==    at 0x1D15206: chunk_callback_t::coro_pool_callback(backfill_queue_entry_t, signal_t*) (backfillee.cc:93)
==23122==    by 0x1D19EE9: coro_pool_t<backfill_queue_entry_t>::worker_run(backfill_queue_entry_t, auto_drainer_t::lock_t) (coro_pool.hpp:68)
==23122==    by 0x1D1D3AD: std::_Mem_fn<void (coro_pool_t<backfill_queue_entry_t>::*)(backfill_queue_entry_t, auto_drainer_t::lock_t)>::operator()(coro_pool_t<backfill_queue_entry_t>*, backfill_queue_entry_t, auto_drainer_t::lock_t) const (in /home/ssd2/grey/rethinkdb/build/debug_valgrind_system/rethinkdb)
==23122==    by 0x1D1D033: _ZNSt5_BindIFSt7_Mem_fnIM11coro_pool_tI22backfill_queue_entry_tEFvS2_N14auto_drainer_t6lock_tEEEPS3_S2_S5_EE6__callIvIEILi0ELi1ELi2EEEET_OSt5tupleIIDpT0_EESt12_Index_tupleIIXspT1_EEE (functional:1146)
==23122==    by 0x1D1CA21: void std::_Bind<std::_Mem_fn<void (coro_pool_t<backfill_queue_entry_t>::*)(backfill_queue_entry_t, auto_drainer_t::lock_t)> ()(coro_pool_t<backfill_queue_entry_t>*, backfill_queue_entry_t, auto_drainer_t::lock_t)>::operator()<, void>() (functional:1205)
==23122==    by 0x1D1C0DB: callable_action_instance_t<std::_Bind<std::_Mem_fn<void (coro_pool_t<backfill_queue_entry_t>::*)(backfill_queue_entry_t, auto_drainer_t::lock_t)> ()(coro_pool_t<backfill_queue_entry_t>*, backfill_queue_entry_t, auto_drainer_t::lock_t)> >::run_action() (callable_action.hpp:31)
==23122==    by 0x1590677: callable_action_wrapper_t::run() (runtime_utils.cc:42)
==23122==    by 0x1580392: coro_t::run() (coroutines.cc:196)
==23122==  Address 0x1d297f20 is 16 bytes inside a block of size 24 free'd
==23122==    at 0x4C2A4BC: operator delete(void*) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==23122==    by 0x167D239: __gnu_cxx::new_allocator<std::_List_node<cond_t*> >::deallocate(std::_List_node<cond_t*>*, unsigned long) (new_allocator.h:98)
==23122==    by 0x167C425: std::_List_base<cond_t*, std::allocator<cond_t*> >::_M_put_node(std::_List_node<cond_t*>*) (stl_list.h:331)
==23122==    by 0x167AE41: std::list<cond_t*, std::allocator<cond_t*> >::_M_erase(std::_List_iterator<cond_t*>) (stl_list.h:1539)
==23122==    by 0x1679D41: std::list<cond_t*, std::allocator<cond_t*> >::pop_front() (stl_list.h:974)
==23122==    by 0x16769B1: log_serializer_t::write_metablock(new_mutex_in_line_t*, signal_t const*, file_account_t*) (log_serializer.cc:608)
==23122==    by 0x1676786: log_serializer_t::index_write_finish(new_mutex_in_line_t*, extent_transaction_t*, file_account_t*) (log_serializer.cc:554)
==23122==    by 0x1676558: log_serializer_t::index_write(new_mutex_in_line_t*, std::vector<index_write_op_t, std::allocator<index_write_op_t> > const&) (log_serializer.cc:524)
==23122==    by 0x16240AD: alt::page_cache_t::do_flush_changes(alt::page_cache_t*, std::map<unsigned long, alt::page_cache_t::block_change_t, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, alt::page_cache_t::block_change_t> > > const&, fifo_enforcer_write_token_t) (page_cache.cc:1310)
==23122==    by 0x1624619: alt::page_cache_t::do_flush_txn_set(alt::page_cache_t*, std::map<unsigned long, alt::page_cache_t::block_change_t, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, alt::page_cache_t::block_change_t> > >*, std::vector<alt::page_txn_t*, std::allocator<alt::page_txn_t*> > const&) (page_cache.cc:1358)
==23122==    by 0x16374D4: _ZNSt5_BindIFPFvPN3alt12page_cache_tEPSt3mapImNS1_14block_change_tESt4lessImESaISt4pairIKmS4_EEERKSt6vectorIPNS0_10page_txn_tESaISF_EEES2_SC_SH_EE6__callIvIEILi0ELi1ELi2EEEET_OSt5tupleIIDpT0_EESt12_Index_tupleIIXspT1_EEE (functional:1146)
==23122==    by 0x16373F9: void std::_Bind<void (*()(alt::page_cache_t*, std::map<unsigned long, alt::page_cache_t::block_change_t, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, alt::page_cache_t::block_change_t> > >*, std::vector<alt::page_txn_t*, std::allocator<alt::page_txn_t*> >))(alt::page_cache_t*, std::map<unsigned long, alt::page_cache_t::block_change_t, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, alt::page_cache_t::block_change_t> > >*, std::vector<alt::page_txn_t*, std::allocator<alt::page_txn_t*> > const&)>::operator()<, void>() (functional:1205)
==23122==
```

**Comment by Tryneus** [comment]
This was caused by a pointer to an entry in a vector being invalidated by a `push_back` on the vector.  The fix is up in review 2490.

**Comment by Tryneus** [comment]
The fix has been approved and merged to `next` in commit b2034ce.  Will be in release 1.16.
